### ISSUE TITLE
The Hand4Whole returns NaNs after first backward.

### ISSUE THREAD

**Issue by xh20** [dqj5182/CONTHO_RELEASE#17]
Dear Daniel,

I followed the instruction for training, and got many NaNs after the first batch without change any hyper parameters in the config file. And found out that the Hand4Whole weights have many NaNs after the first backward, especially the self.obj_id_emded.weight. The hand4whole module used in the CONTHO model has an additional object embedding layer (in PoseResNet), which is different to the original one. I have some following questions:
1. How did you obtain the checkpoint for the hand4whole module? Did you pretrain it or download it somewhere?
2. Did you also put it in the trainable_modules list, even it is initialized with checkpoints? I guess the checkpoint is already close to the optimum, and the global learning rate is probably too large for it.

Thank you for your support! Very Inspiring project!

Best,
Hao

**Comment by dqj5182** [comment]
Thank you for your interest in our work!

Regarding your error on `self.obj_id_embed.weight` and error on Hand4Whole weights, I am speculating that you are training CONTHO on your own dataset. I think this problem is currently not happening in both BEHAVE and InterCap training. Could you double-check the quality of your ground-truth 3D human mesh and whether it is in same format with ones that we used for training (camera coordinate)? Also, for `self.obj_id_embed.weight`, I think you may not have set the number of objects for CONTHO. CONTHO requires specific setting for number of object you are trying to train/inference. Please check these criteria and let me know if the problem still exists.

1. The checkpoint for the hand4whole module is from [official Hand4Whole repository](https://github.com/mks0601/Hand4Whole_RELEASE/tree/main?tab=readme-ov-file) (Download the pre-trained Hand4Whole from [here](https://drive.google.com/file/d/1r0LfI_ATI8NmhOj4y8eqLlEDdNQj5SCV/view?usp=sharing). We have not pre-trained the Hand4Whole module before training CONTHO. I have just double-checked and they are exactly identical in terms of model weight values.
2. Yes. We also train Hand4Whole along with other modules in CONTHO. The purpose of such learning strategy is to train 3D human mesh reconstruction module (i.e. Hand4Whole) on human-object interaction scenarios. Such scenarios contain multiple occlusions by objects, which may not have been trained well on the base model (i.e., Hand4Whole). If your circumstance requires better generalization than occlusion handling, then I think not training 3D human mesh reconstruction module might be more beneficial. But if you try to train CONTHO on specific human-object interaction scenarios, then our default training method should suffice.

Sincerely,
Daniel

**Comment by xh20** [comment]
Hi Daniel,

Thank you for your reply. It works now. I am using the Behave dataset as well. The number of objects is 20. The differences are number of GPUs and batch size, I used two GPUs and 32 bs, which ends up with NaN in the Hand4Whole or SVD unsolvable in the rigid_transform_3D. When I change it to 1 GPU and 16 bs, it works perfectly. I think you can close this issue. I will open another issue, when I find the parallel computing problems.
Thank you again for your effort.

Best,
Hao

**Comment by dqj5182** [comment]
Glad to hear that you have find a way to solve the problem!

May I ask if you could try this [pull request](https://github.com/dqj5182/CONTHO_RELEASE/pull/13)? I usually don't take pull requests, but if the current code is not trainable in multi-gpu environment at all and this pull request works in that circumstance, I am considering of accepting this pull request.

Hope to hear from you soon.
